ksql-cli/src/main/java/io/confluent/ksql/cli/console/Console.java

```diff
@@ -135,69 +139,120 @@
+      final Console theConsole = console.get();
+      return theConsole != null && theConsole.isCliCommand(line);
+    };
 
-    this.objectMapper = new ObjectMapper().disable(JsonGenerator.Feature.AUTO_CLOSE_TARGET);
+    final Path historyFilePath = Paths.get(System.getProperty(
+        "history-file",
+        System.getProperty("user.home")
+            + "/.ksql-history"
+    )).toAbsolutePath();
+
+    final KsqlTerminal terminal = new JLineTerminal(isCliCommand, historyFilePath);
 
     final Supplier<String> versionSuppler =
         () -> restClient.getServerInfo().getResponse().getVersion();
-    CliCommandRegisterUtil.registerDefaultCommands(this, versionSuppler);
-  }
 
-  public abstract PrintWriter writer();
+    return new Console(outputFormat, versionSuppler, terminal, new NoOpRowCaptor());
+  }
 
-  public abstract void flush();
+  public Console(
+      final OutputFormat outputFormat,
+      final Supplier<String> versionSuppler,
+      final KsqlTerminal terminal,
+      final RowCaptor rowCaptor
+  ) {
+    this.outputFormat = Objects.requireNonNull(outputFormat, "outputFormat");
+    this.terminal = Objects.requireNonNull(terminal, "terminal");
+    this.rowCaptor = Objects.requireNonNull(rowCaptor, "rowCaptor");
+    this.cliSpecificCommands = Maps.newLinkedHashMap();
+    this.objectMapper = new ObjectMapper().disable(JsonGenerator.Feature.AUTO_CLOSE_TARGET);
 
-  public abstract int getWidth();
+    CliCommandRegisterUtil.registerDefaultCommands(this, versionSuppler);
+  }
 
-  /* jline specific */
+  public PrintWriter writer() {
+    return terminal.writer();
+  }
 
-  protected abstract LineReader buildLineReader();
+  public void flush() {
+    terminal.flush();
+  }
 
-  public abstract void clearScreen();
+  public int getWidth() {
+    return terminal.getWidth();
+  }
 
-  public abstract void handle(Signal signal, SignalHandler signalHandler);
+  public void clearScreen() {
+    terminal.clearScreen();
+  }
 
-  /* public */
+  public void handle(final Signal signal, final SignalHandler signalHandler) {
+    terminal.handle(signal, signalHandler);
+  }
 
-  public void addResult(final GenericRow row) {
-    // do nothing by default, test classes can use this method to obtain typed results
+  @Override
+  public void close() {
+    terminal.close();
   }
 
-  public void addResult(final List<String> columnHeaders, final List<List<String>> rowValues) {
-    // do nothing by default, test classes can use this method to obtain typed results
+  public void addResult(final List<List<String>> rowValues) {
+    rowCaptor.addRows(rowValues);
   }
 
   public Map<String, CliSpecificCommand> getCliSpecificCommands() {
     return cliSpecificCommands;
   }
 
-  public LineReader getLineReader() {
-    if (lineReader == null) {
-      lineReader = buildLineReader();
+  public String readLine() {
+    String line;
+
+    do {
+      line = terminal.readLine();
+
+    } while (!maybeHandleCliSpecificCommands(line));
```

Comment:
Should this be while (maybeHandleCliSpecificCommand(line))? We want to break the loop when we have a line that's not a cli-specific command.